### PR TITLE
Overwrite

### PR DESCRIPTION
**Pull request by jhfjhfj1**
### Which issue(s) does this Pull Request fix?

resolves #000

### Details of the Pull Request

1. Add the `overwrite` arg to all APIs and to AutoTuner.
2. Refactored the saving mechanism for the hypermodels to use `config` and `weights`.
3. HyperGraph will not only save the configs of each block but also the graph structure.

### CODE REVIEW

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/autokeras/autokeras/pullRequest?prid=4564373&bid=15345713)

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 9
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)

File: autokeras/encoder.py [hunk 1]
```diff
@@ -68,14 +73,14 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
         self._label_to_vec = {}
 
-    def get_state(self):
-        state = super().get_state()
-        state.update({'label_to_vec': self._label_to_vec})
-        return state
+    def get_weights(self):
```
Comment (omalleyt12) [inline autokeras/encoder.py]: Why the change from `state` to `weights`? It looks like this method is returning more than just weights

File: autokeras/hypermodel/base.py [hunk 2]
```diff
@@ -1,13 +1,95 @@
...
 import pandas as pd
 import tensorflow as tf
+from kerastuner.engine import stateful
 from tensorflow.python.util import nest
 
 from autokeras import utils
 
 
-class Node(kerastuner.engine.stateful.Stateful):
+class Pickable(stateful.Stateful):
```
Comment (omalleyt12) [inline autokeras/hypermodel/base.py]: Nit: Is this supposed to be `Picklable`?

File: autokeras/hypermodel/graph.py [hunk 3]
```diff
@@ -463,4 +476,69 @@ def hyper_build(self, hp):
...
         for output_node in self.outputs:
             outputs.append(old_node_to_new[output_node])
-        return PlainGraph(inputs, outputs, override_hps=self.override_hps)
+        return PlainGraph(inputs=inputs,
+                          outputs=outputs,
+                          override_hps=self.override_hps)
+
+    def get_config(self):
+        config = super().get_config()
+        block_classes = [block.__class__.__name__ for block in self.blocks]
```
Comment (omalleyt12) [inline autokeras/hypermodel/graph.py]: Can blocks have arguments? If so, should this be `[block.get_config() for block in self.blocks]`

File: autokeras/hypermodel/graph.py [hunk 4]
```diff
@@ -463,4 +476,69 @@ def hyper_build(self, hp):
...
+        node_classes = [node.__class__.__name__ for node in self._nodes]
```
Comment (omalleyt12) [inline autokeras/hypermodel/graph.py]: Same here

File: autokeras/hypermodel/base.py [hunk 5]
```diff
@@ -1,13 +1,95 @@
...
+
+    def get_config(self):
+        """Returns the current config of this object.
+
+        # Returns
+            Dictionary.
+        """
+        raise NotImplementedError
+
+    def set_config(self, config):
```
Comment (omalleyt12) [inline autokeras/hypermodel/base.py]: This should probably follow the Keras convention:

```python
@classmethod
def from_config(self, config):
  ...
```

"State" and "config" should probably be entirely separate (we don't make this distinction very well in Keras Tuner at the moment, but AutoKeras can do better I think)

The methods should probably be:

```get_config``` -> Returns the arguments the object was constructed with
```from_config``` -> Makes the object
```get_state``` -> Returns non-config related state. I.e. only things that changed from instantiation. This is the equivalent `keras.layers.Layer.get_weights()`
```set_state`` -> Sets the non-config relate state. This is the equivalent of `keras.layers.Layer.set_weights()`

This way you can reuse the Keras mechanisms for `get_config`, `from_config`, and `custom_objects` in order to support user-provided objects

`save` and `reload` could make use of the other methods, but `get_weights` and `set_weights` would be redundant

File: autokeras/hypermodel/graph.py [hunk 6]
```diff
@@ -463,4 +476,69 @@ def hyper_build(self, hp):
...
+            'node_classes': node_classes,  # List of strings of class names.
+            'block_inputs': block_inputs,  # Dict {id: List of node_ids}.
+            'block_outputs': block_outputs,  # Dict {id: List of node_ids}.
+            'node_inputs': node_inputs,  # Dict {id: List of block_ids}.
+            'node_outputs': node_outputs,  # Dict {id: List of block_ids}.
+        })
+        return config
+
+    def set_config(self, config):
+        self.blocks = [getattr(autokeras, block_class)()
```
Comment (omalleyt12) [inline autokeras/hypermodel/graph.py]: I expect you'll want to support user-defined blocks

To do so, if blocks have a `from_config` like described above, then you can make use of Keras's ` custom_object` mechanisms:

https://github.com/tensorflow/tensorflow/blob/649d1b34da7a90f0dbc024911751a34bef591290/tensorflow/python/keras/utils/generic_utils.py#L42

```python
@classmethod
def from_config(self, config, custom_objects=None):
  self.blocks = [
     tf.keras.deserialize_keras_object(block_config, custom_objects=custom_objects)
    for block_config in config['blocks']]
...
```

Comment (omalleyt12) [reply autokeras/hypermodel/graph.py]: Same for nodes below

**Comment by jhfjhfj1** [comment]
@omalleyt12 Thank you for the review! Great suggestions. I will try to follow the get from config and get set state style.

File: autokeras/auto_model.py [hunk 7]
```diff
@@ -44,6 +46,7 @@ def __init__(self,
                  directory=None,
                  objective='val_loss',
                  tuner='greedy',
+                 overwrite=True,
```
[the remark this answers is not in the document]
Comment (jhfjhfj1) [reply autokeras/auto_model.py]: Cool. I can change it to default to False.

File: autokeras/hypermodel/graph.py (hunk 6 above)
Comment (jhfjhfj1) [reply autokeras/hypermodel/graph.py]: To my understanding, all the blocks in AutoKeras are custom_objects to Keras. So we need to union the AutoKeras objects dict and the custom_objects passed by the user as the `custom_objects` to the `deserialize_keras_object` function, right?

Comment (omalleyt12) [reply autokeras/hypermodel/graph.py]: That's one approach, another way is to do a module-level approach like what Keras does, which is that for each object type, there is a `deserialize` method (for example, `tf.keras.metrics.deserialize(config, custom_objects=None)`):

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L3214

This function collects that particular module's `globals()` and uses those as the deserialization names and objects dict.

So you could have something like `autokeras.blocks.deserialize(config, custom_objects=None)` that hooks into `deserialize_keras_object` in the same way. Then you would use that when you know the thing you're deserializing is a Block

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 15
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)

**Comment by jhfjhfj1** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/jhfjhfj1/autokeras/pullRequest?prid=4564373&bid=15345713)
